test/common/http/conn_manager_impl_test.cc

```diff
@@ -3359,6 +3359,59 @@ TEST_F(HttpConnectionManagerImplTest, FilterClearRouteCache) {
   conn_manager_->onData(fake_input, false);
 }
 
+// Dummy PerConnectionSharedPtr creator-function; this creator-function can be as robust
+// and state-preserving as the inheriting interface desires.
+PerConnectionObjectSharedPtr test_creation_function(MutableHttpConnection&) {
+  return std::make_shared<PerConnectionObject>();
+}
+
+TEST_F(HttpConnectionManagerImplTest, MutableHttpConnection) {
+  setup(false, "", true, true);
+
+  MutableHttpConnection mutable_connection;
+  Http::StreamFilterCallbacks::PerConnectionObjectCreator creator_function = test_creation_function;
+  // Types of per connection objects
+  const std::string test_str_obj1 = "read_buffer_handle";
+  const std::string test_str_obj2 = "recv_lowat_handle";
+
+  setupFilterChain(1, 0); // Create filter chain
+
+  EXPECT_CALL(*static_cast<const Router::MockScopedConfig*>(
+                  scopedRouteConfigProvider()->config<Router::ScopedConfig>().get()),
+              getRouteConfig(_))
+      .Times(2)
+      .WillRepeatedly(Return(nullptr));
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> void {
+    StreamDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+    HeaderMapPtr headers{
+        new TestHeaderMapImpl{{":authority", "host"}, {":method", "GET"}, {":path", "/foo"}}};
+    decoder->decodeHeaders(std::move(headers), true);
+    data.drain(4);
+  }));
+
+  EXPECT_CALL(*decoder_filters_[0], decodeHeaders(_, true))
```

Comment:
To be a robust test, you should have at least 2 filters and verify that you get the same object across filters, etc.

Reply:
Thanks for the comments. I have also added a nullptr check for `creation_function` and empty `object_name`